packages/core/core/src/BundleGraph.js

```diff
@@ -1,70 +1,354 @@
 // @flow strict-local
 
-import type {GraphTraversalCallback} from '@parcel/types';
+import type {
+  BundleGroup,
+  Dependency as IDependency,
+  GraphVisitor,
+  Symbol,
+  SymbolResolution,
+  TraversalActions
+} from '@parcel/types';
+
+import type {AssetNode, Bundle, BundleGraphNode, DependencyNode} from './types';
 import type Asset from './Asset';
-import type {Bundle, BundleGraphNode} from './types';
+import type Graph from './Graph';
 
-import Graph, {type GraphOpts} from './Graph';
+import invariant from 'assert';
+import crypto from 'crypto';
+import nullthrows from 'nullthrows';
 
-export default class BundleGraph extends Graph<BundleGraphNode> {
-  constructor(opts?: GraphOpts<BundleGraphNode>) {
-    super(opts);
-    this.setRootNode({
-      type: 'root',
-      id: 'root',
-      value: null
-    });
+import {getBundleGroupId} from './utils';
+import {mapVisitor} from './Graph';
+
+type BundleGraphEdgeTypes = 'contains' | 'bundle' | 'references';
+
+export default class BundleGraph {
+  _graph: Graph<BundleGraphNode, BundleGraphEdgeTypes>;
```

Comment:
Unlike `AssetGraph`, `RequestGraph`, etc., this *has* a graph rather than *is* a graph. It `_` since I'd like to keep the internal graph structures as details of these kinds of classes.

That said, that detail is ignored a lot in this PR and external things frequently use `._graph`, mostly for convenience. Arguably we could create a higher-level api rather than checking node membership, connections, etc. But maybe the practical thing to do is make this extend Graph like the others?